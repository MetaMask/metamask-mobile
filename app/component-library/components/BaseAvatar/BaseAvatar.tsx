/* eslint-disable react/prop-types */
import React from 'react';
import { View } from 'react-native';
import { useStyles } from '../../hooks/useStyles';
import { BaseAvatarProps } from './BaseAvatar.types';
import styleSheet from './BaseAvatar.styles';

const BaseAvatar: React.FC<BaseAvatarProps> = ({ size, style, children }) => {
  const { styles } = useStyles(styleSheet, {
    size,
    style,
  });

  return <View style={styles.container}>{children}</View>;
};

export default BaseAvatar;
