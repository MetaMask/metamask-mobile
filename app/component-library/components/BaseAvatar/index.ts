export { default } from './BaseAvatar';
export { BaseAvatarSize } from './BaseAvatar.types';
