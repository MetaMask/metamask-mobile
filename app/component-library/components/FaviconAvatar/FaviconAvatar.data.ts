/* eslint-disable import/prefer-default-export */
export const testImageUrl = 'https://uniswap.org/favicon.ico';
