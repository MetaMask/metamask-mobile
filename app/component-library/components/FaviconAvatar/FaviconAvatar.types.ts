import { ImageStyle, ViewStyle } from 'react-native';
import {
  BaseAvatarProps,
  BaseAvatarSize,
} from '../BaseAvatar/BaseAvatar.types';
import { IconSize } from '../Icon';

/**
 * FaviconAvatar component props.
 */
export interface FaviconAvatarProps extends BaseAvatarProps {
  /**
   * An icon URL
   */
  imageUrl: string;
}

/**
 * Favicon component style sheet.
 */
export interface FaviconAvatarStyleSheet {
  base: ViewStyle;
  image: ImageStyle;
}

/**
 * Style sheet input parameters.
 */
export interface FaviconAvatarStyleSheetVars
  extends Pick<FaviconAvatarProps, 'style'> {
  error: any;
}

/**
 * Mapping of IconSize by BaseAvatarSize.
 */
export type IconSizeByAvatarSize = {
  [key in BaseAvatarSize]: IconSize;
};
