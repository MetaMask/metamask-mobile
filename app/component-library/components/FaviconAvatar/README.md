# FaviconAvatar

FaviconAvatar is an Image component that renders an icon that is provided in the form of a URL. This component is based on the [BaseAvatar](../BaseAvatar/BaseAvatar.tsx) component.

## Props

This component extends [BaseAvatarProps](../BaseAvatar/BaseAvatar.types.ts#L17) from `BaseAvatar` component.

### `size`

Enum to select between size variants.

| <span style="color:gray;font-size:14px">TYPE</span>    | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :----------------------------------------------------- | :------------------------------------------------------ |
| [BaseAvatarSize](../BaseAvatar/BaseAvatar.types.ts#L6) | Yes                                                     |

### `imageUrl`

An icon URL.

| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :-------------------------------------------------- | :------------------------------------------------------ |
| String                                              | Yes                                                     |
