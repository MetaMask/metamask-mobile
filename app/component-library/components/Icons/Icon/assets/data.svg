<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none"><path d="M20.281 14.25a2.25 2.25 0 1 0 0-4.5 2.25 2.25 0 0 0 0 4.5zM20.281 6.25a2.25 2.25 0 1 0 0-4.5 2.25 2.25 0 0 0 0 4.5zM20.281 22.25a2.25 2.25 0 1 0 0-4.5 2.25 2.25 0 0 0 0 4.5zM4.281 14.25a2.25 2.25 0 1 0 0-4.5 2.25 2.25 0 0 0 0 4.5z" fill="currentColor"/><path d="M19.281 12.75c.41 0 .75-.34.75-.75s-.34-.75-.75-.75h-7.25V7c0-1.58.67-2.25 2.25-2.25h5c.41 0 .75-.34.75-.75s-.34-.75-.75-.75h-5c-2.42 0-3.75 1.33-3.75 3.75v4.25h-5.25c-.41 0-.75.34-.75.75s.34.75.75.75h5.25V17c0 2.42 1.33 3.75 3.75 3.75h5c.41 0 .75-.34.75-.75s-.34-.75-.75-.75h-5c-1.58 0-2.25-.67-2.25-2.25v-4.25h7.25z" fill="currentColor"/></svg>