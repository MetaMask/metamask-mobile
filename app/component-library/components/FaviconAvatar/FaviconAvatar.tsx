/* eslint-disable react/prop-types */
import React, { useCallback, useState } from 'react';
import { Image, ImageErrorEventData, NativeSyntheticEvent } from 'react-native';
import { useStyles } from '../../../component-library/hooks';
import BaseAvatar, { BaseAvatarSize } from '../BaseAvatar';
import {
  FaviconAvatarProps,
  IconSizeByAvatarSize,
} from './FaviconAvatar.types';
import stylesheet from './FaviconAvatar.styles';
import Icon, { IconName, IconSize } from '../Icon';
import { FAVICON_AVATAR_IMAGE_ID } from '../../../constants/test-ids';

const iconSizeByAvatarSize: IconSizeByAvatarSize = {
  [BaseAvatarSize.Xs]: IconSize.Xs,
  [BaseAvatarSize.Sm]: IconSize.Sm,
  [BaseAvatarSize.Md]: IconSize.Md,
  [BaseAvatarSize.Lg]: IconSize.Lg,
  [BaseAvatarSize.Xl]: IconSize.Xl,
};

const FaviconAvatar: React.FC<FaviconAvatarProps> = ({
  imageUrl,
  size,
  style,
}) => {
  const [error, setError] = useState(undefined);
  const { styles } = useStyles(stylesheet, { style, error });

  const onError = useCallback(
    (e: NativeSyntheticEvent<ImageErrorEventData>) =>
      setError(e.nativeEvent.error),
    [setError],
  );

  return (
    <BaseAvatar size={size} style={styles.base}>
      {error ? (
        <Icon size={iconSizeByAvatarSize[size]} name={IconName.GlobalFilled} />
      ) : (
        <Image
          testID={FAVICON_AVATAR_IMAGE_ID}
          source={{ uri: imageUrl }}
          style={styles.image}
          resizeMode={'contain'}
          onError={onError}
        />
      )}
    </BaseAvatar>
  );
};

export default FaviconAvatar;
