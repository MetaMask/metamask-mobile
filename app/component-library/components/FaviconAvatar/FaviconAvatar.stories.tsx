import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { select } from '@storybook/addon-knobs';
import { BaseAvatarSize } from '../BaseAvatar';
import FaviconAvatar from '.';
import { testImageUrl } from './FaviconAvatar.data';

storiesOf('Component Library / FaviconAvatar', module)
  .addDecorator((getStory) => getStory())
  .add('Default', () => {
    const sizeSelector = select(
      'size',
      BaseAvatarSize,
      BaseAvatarSize.Md,
      'Avatar Size',
    );

    return <FaviconAvatar size={sizeSelector} imageUrl={testImageUrl} />;
  })
  .add('With Error', () => {
    const sizeSelector = select(
      'size',
      BaseAvatarSize,
      BaseAvatarSize.Md,
      'Avatar Size',
    );

    return <FaviconAvatar size={sizeSelector} imageUrl={''} />;
  });
