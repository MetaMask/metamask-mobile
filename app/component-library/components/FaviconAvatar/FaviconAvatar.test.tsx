import React from 'react';
import { shallow } from 'enzyme';
import { BaseAvatarSize } from '../BaseAvatar';
import FaviconAvatar from '.';
import { testImageUrl } from './FaviconAvatar.data';
import { FAVICON_AVATAR_IMAGE_ID } from '../../../constants/test-ids';

describe('FaviconAvatar', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(
      <FaviconAvatar size={BaseAvatarSize.Xl} imageUrl={testImageUrl} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render favicon', () => {
    const wrapper = shallow(
      <FaviconAvatar size={BaseAvatarSize.Xl} imageUrl={testImageUrl} />,
    );
    const imageComponent = wrapper.findWhere(
      (node) => node.prop('testID') === FAVICON_AVATAR_IMAGE_ID,
    );
    expect(imageComponent.exists()).toBe(true);
  });

  it('should render fallback', () => {
    const wrapper = shallow(
      <FaviconAvatar size={BaseAvatarSize.Xl} imageUrl={testImageUrl} />,
    );
    const prevImageComponent = wrapper.findWhere(
      (node) => node.prop('testID') === FAVICON_AVATAR_IMAGE_ID,
    );
    // Simulate onError on Image component
    prevImageComponent.props().onError({ nativeEvent: { error: 'ERROR!' } });
    const currentImageComponent = wrapper.findWhere(
      (node) => node.prop('testID') === FAVICON_AVATAR_IMAGE_ID,
    );
    expect(currentImageComponent.exists()).toBe(false);
  });
});
