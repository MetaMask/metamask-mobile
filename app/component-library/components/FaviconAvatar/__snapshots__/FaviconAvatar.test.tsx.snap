// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FaviconAvatar should match the snapshot 1`] = `
<BaseAvatar
  size="48"
  style={
    Object {
      "borderRadius": 0,
    }
  }
>
  <Image
    onError={[Function]}
    resizeMode="contain"
    source={
      Object {
        "uri": "https://uniswap.org/favicon.ico",
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
    testID="favicon-avatar-image"
  />
</BaseAvatar>
`;
